File: app/services/ai/idea_generator.rb
# frozen_string_literal: true

module Ai
  class IdeaGenerator < BaseAnalyzer
    SYSTEM_PROMPT = <<~PROMPT
      You are an expert product strategist generating solution ideas for identified insights.
      For each insight, propose actionable ideas that could address the underlying issue or opportunity.

      For each idea:
      1. Create a clear, actionable title
      2. Write a detailed description of the proposed solution
      3. Classify the type: quick_win, feature, improvement, process_change, or investigation
      4. Estimate effort: trivial, small, medium, large, or extra_large
      5. Estimate impact: minimal, low, moderate, high, or transformational
      6. Provide your confidence score (0-100)
      7. Explain the rationale for this idea
      8. List potential risks or considerations
      9. Provide implementation hints or first steps

      Respond in JSON format:
      {
        "ideas": [
          {
            "title": "Clear idea title",
            "description": "Detailed solution description",
            "idea_type": "quick_win|feature|improvement|process_change|investigation",
            "effort_estimate": "trivial|small|medium|large|extra_large",
            "impact_estimate": "minimal|low|moderate|high|transformational",
            "confidence_score": 75,
            "rationale": "Why this solution makes sense",
            "risks": "Potential risks or considerations",
            "implementation_hints": ["First step", "Second step"]
          }
        ]
      }

      Guidelines:
      - Generate 1-3 ideas per insight
      - Prioritize ideas with high impact and low effort (quick wins)
      - Be specific and actionable
      - Consider both technical and process solutions
      - Include at least one quick win if applicable
    PROMPT

    def generate(insight)
      result = analyze_insight(insight)

      if result[:error]
        Rails.logger.error("Idea generation failed for insight ##{insight.id}: #{result[:error]}")
        return { ideas: [], created: 0, error: result[:error] }
      end

      created = create_ideas_from_result(result, insight)

      { ideas: created, created: created.count }
    end

    def generate_batch(insights)
      all_ideas = []

      insights.each do |insight|
        result = generate(insight)
        all_ideas.concat(result[:ideas]) unless result[:error]
        rate_limit_sleep
      end

      { ideas: all_ideas, created: all_ideas.count }
    end

    private

    def analyze_insight(insight)
      prompt = build_insight_prompt(insight)
      call_claude(prompt, system_prompt: SYSTEM_PROMPT)
    end

    def build_insight_prompt(insight)
      parts = ["Generate solution ideas for the following insight:\n"]
      parts << "Title: #{insight.title}"
      parts << "Description: #{insight.description}"
      parts << "Type: #{insight.insight_type}"
      parts << "Severity: #{insight.severity}"
      parts << "Affected Users: #{insight.affected_users_count}"
      parts << "Feedback Count: #{insight.feedback_count}"

      if insight.evidence.present?
        parts << "\nSupporting Evidence:"
        insight.evidence.each { |e| parts << "- #{e}" }
      end

      parts.join("\n")
    end

    def create_ideas_from_result(result, insight)
      created_ideas = []

      ideas_data = result[:ideas] || []
      ideas_data.each do |idea_data|
        idea = create_idea(idea_data)
        next unless idea

        link_idea_to_insight(idea, insight)
        created_ideas << idea
      end

      created_ideas
    end

    def create_idea(data)
      Idea.create!(
        title: data[:title],
        description: data[:description],
        idea_type: data[:idea_type],
        effort_estimate: data[:effort_estimate],
        impact_estimate: data[:impact_estimate],
        confidence_score: data[:confidence_score] || 0,
        rationale: data[:rationale],
        risks: data[:risks],
        implementation_hints: data[:implementation_hints] || [],
        pm_persona: @pm_persona,
        status: :proposed
      )
    rescue ActiveRecord::RecordInvalid => e
      Rails.logger.error("Failed to create idea: #{e.message}")
      nil
    end

    def link_idea_to_insight(idea, insight)
      IdeaInsight.find_or_create_by!(
        idea: idea,
        insight: insight
      ) do |ii|
        ii.address_level = calculate_address_level(idea)
        ii.address_description = "Generated solution for #{insight.title}"
      end
    end

    def calculate_address_level(idea)
      impact_score = Idea.impact_estimates[idea.impact_estimate] || 0
      case impact_score
      when 4 then 4
      when 3 then 3
      when 2 then 2
      else 1
      end
    end
  end
end

File: apps/worker/src/ai/idea-generator.ts
/**
 * AI idea generation — ports Rails IdeaGenerator.
 * Creates actionable solution ideas from insights.
 */
import { eq, desc } from "drizzle-orm";
import type { Database } from "@customer-pulse/db/client";
import { insights, ideas, ideaInsights } from "@customer-pulse/db/client";
import { callClaudeJson } from "./call-claude.js";

interface GeneratedIdea {
  title: string;
  description: string;
  type: string;
  effort_estimate: string;
  impact_estimate: string;
  confidence_score: number;
  rationale: string;
  risks: string;
  implementation_hints: string[];
  address_level: number;
}

const TYPE_MAP: Record<string, number> = { quick_win: 0, feature: 1, improvement: 2, process_change: 3, investigation: 4 };
const EFFORT_MAP: Record<string, number> = { trivial: 0, small: 1, medium: 2, large: 3, extra_large: 4 };
const IMPACT_MAP: Record<string, number> = { minimal: 0, low: 1, moderate: 2, high: 3, transformational: 4 };

const SYSTEM_PROMPT = `You are a product strategist. Given an insight about customer feedback, generate 1-3 actionable solution ideas.

For each idea, return:
- "title": concise title (< 100 chars)
- "description": 2-3 sentence description
- "type": one of "quick_win", "feature", "improvement", "process_change", "investigation"
- "effort_estimate": one of "trivial", "small", "medium", "large", "extra_large"
- "impact_estimate": one of "minimal", "low", "moderate", "high", "transformational"
- "confidence_score": 0-100 how confident you are this would work
- "rationale": why this idea addresses the insight
- "risks": potential risks or downsides
- "implementation_hints": array of specific implementation steps or considerations
- "address_level": 0-4 how fully this idea addresses the insight (0=barely, 4=completely)

Prioritize quick wins (high impact, low effort) when possible. Return a JSON array.
Respond with ONLY the JSON array, no markdown fences.`;

export async function generateIdeasForInsight(
  db: Database,
  insightId: number,
): Promise<number> {
  const [insight] = await db.select().from(insights).where(eq(insights.id, insightId)).limit(1);
  if (!insight) return 0;

  const severityNames: Record<number, string> = { 0: "informational", 1: "minor", 2: "moderate", 3: "major", 4: "critical" };
  const typeNames: Record<number, string> = { 0: "problem", 1: "opportunity", 2: "trend", 3: "risk", 4: "user_need" };

  const result = await callClaudeJson<GeneratedIdea[]>({
    system: SYSTEM_PROMPT,
    user: `Insight: "${insight.title}"\nType: ${typeNames[insight.insightType]}\nSeverity: ${severityNames[insight.severity]}\nAffected users: ~${insight.affectedUsersCount}\n\n${insight.description}`,
    maxTokens: 4096,
  });

  if (!result || !Array.isArray(result)) return 0;

  const now = new Date();
  let created = 0;

  for (const item of result) {
    const [newIdea] = await db.insert(ideas).values({
      projectId: insight.projectId,
      title: item.title ?? "Untitled Idea",
      description: item.description ?? "",
      ideaType: TYPE_MAP[item.type] ?? 0,
      effortEstimate: EFFORT_MAP[item.effort_estimate] ?? 2,
      impactEstimate: IMPACT_MAP[item.impact_estimate] ?? 2,
      confidenceScore: item.confidence_score ?? 50,
      status: 0, // proposed
      rationale: item.rationale ?? null,
      risks: item.risks ?? null,
      implementationHints: (item.implementation_hints ?? []) as unknown[],
      metadata: {},
      createdAt: now,
      updatedAt: now,
    }).returning({ id: ideas.id });

    if (newIdea) {
      try {
        await db.insert(ideaInsights).values({
          ideaId: newIdea.id,
          insightId: insight.id,
          addressLevel: item.address_level ?? 2,
          addressDescription: item.rationale ?? null,
          createdAt: now,
          updatedAt: now,
        });
      } catch { /* duplicate — ignore */ }
      created++;
    }
  }

  return created;
}

export async function generateIdeasForProject(
  db: Database,
  projectId: number,
): Promise<number> {
  // Get recent insights that don't have ideas yet
  const recentInsights = await db
    .select({ id: insights.id })
    .from(insights)
    .where(eq(insights.projectId, projectId))
    .orderBy(desc(insights.createdAt))
    .limit(20);

  let total = 0;
  for (const insight of recentInsights) {
    const count = await generateIdeasForInsight(db, insight.id);
    total += count;
  }
  return total;
}
